### ISSUE TITLE
NullableWalker's result recording for patterns needs to be updated

### ISSUE THREAD

**Issue by 333fred** [dotnet/roslyn#35032]
Currently patterns aren't supported by the result recording steps, this should be updated.

**Comment by gafter** [comment]
I don't think patterns "have" nullability.  What would that mean?

**Comment by 333fred** [comment]
There are a few things that I mean here:

1. Declaration patterns need to have the declared variable nullability recorded.
2. In error scenarios, we can have constant patterns with non-constant patterns. Those should either be "analyzed" to record information or should be set to not null, as we deem appropriate.
3. In general, we should go through what information can be retrieved from the various pattern structures and determine what nullabilities those should have, if any at all.

**Comment by gafter** [comment]
Declaration patterns do not permit a nullable annotation.

**Comment by 333fred** [comment]
@gafter I agree with that in correct code, but we need to verify what happens if I were to type `x is string? s`.

**Comment by gafter** [comment]
You will get a syntax error that the : of your ?: expression is missing.

-Neal
________________________________
